File: app/src/main/java/org/andstatus/app/net/social/InputPage.java
/*
 * Copyright (C) 2020 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.andstatus.app.net.social;

import org.andstatus.app.util.HasEmpty;

import java.util.List;

public abstract class InputPage<T> implements HasEmpty<T> {
    public final TimelinePosition firstPosition;
    public final TimelinePosition youngerPosition;
    public final TimelinePosition thisPosition;
    public final TimelinePosition olderPosition;
    public final List<T> activities;
    private final boolean isEmpty;

    protected InputPage(AJsonCollection jsonCollection, List<T> activities) {
        firstPosition = TimelinePosition.of(jsonCollection.firstPage.getId());
        youngerPosition = TimelinePosition.of(jsonCollection.prevPage.getId());
        thisPosition = TimelinePosition.of(jsonCollection.getId());
        olderPosition = TimelinePosition.of(jsonCollection.nextPage.getId());
        this.activities = activities;
        isEmpty = jsonCollection.isEmpty();
    }

    public boolean isEmpty() {
        return isEmpty;
    }

    public T get(int ind) {
        return ind < 0 || ind > activities.size()
                ? empty()
                : activities.get(ind);
    }

    public int size() {
        return activities.size();
    }
}

File: app/src/main/java/org/andstatus/app/net/social/InputPage.kt
/*
 * Copyright (C) 2020 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.andstatus.app.net.social

import org.andstatus.app.net.social.TimelinePosition
import org.andstatus.app.util.HasEmpty

abstract class InputPage<T> protected constructor(jsonCollection: AJsonCollection,
                                                  val items: MutableList<T>) : HasEmpty<T> {
    val firstPosition: TimelinePosition = TimelinePosition.Companion.of(jsonCollection.firstPage.getId())
    val youngerPosition: TimelinePosition = TimelinePosition.Companion.of(jsonCollection.prevPage.getId())
    val thisPosition: TimelinePosition = TimelinePosition.Companion.of(jsonCollection.getId())
    val olderPosition: TimelinePosition = TimelinePosition.Companion.of(jsonCollection.nextPage.getId())

    override val isEmpty: Boolean = jsonCollection.isEmpty

    operator fun get(ind: Int): T? {
        return if (ind < 0 || ind > items.size) empty() else items.get(ind)
    }

    fun size(): Int {
        return items.size
    }

    override fun toString(): String {
        return items.toString()
    }
                                                  }